source/grid/grid_generator.cc

```diff
@@ -5105,6 +5106,357 @@ namespace GridGenerator
 
 
 
+  template <>
+  void
+  tee(Triangulation<3, 3> &                             tria,
+      const std::array<std::pair<Point<3>, double>, 3> &openings,
+      const std::pair<Point<3>, double> &               bifurcation)
+  {
+    constexpr unsigned int dim      = 3;
+    constexpr unsigned int spacedim = 3;
+    using vector                    = Tensor<1, spacedim, double>;
+
+    constexpr unsigned int n_pipes   = 3;
+    constexpr double       tolerance = 1.e-12;
+
+#  ifdef DEBUG
+    // Verify user input.
+    Assert(bifurcation.second > 0,
+           ExcMessage("Invalid input: negative radius."));
+    for (const auto &opening : openings)
+      Assert(opening.second > 0, ExcMessage("Invalid input: negative radius."));
+#  endif
+
+    // Each pipe segment will be identified by the index of its opening in the
+    // parameter array. To determine the next and previous entry in the array
+    // for a given index, we create auxiliary functions.
+    const auto cyclic = [=](const unsigned int i) -> unsigned int {
+      return (i < (n_pipes - 1)) ? i + 1 : 0;
+    };
+    const auto anticyclic = [=](const unsigned int i) -> unsigned int {
+      return (i > 0) ? i - 1 : n_pipes - 1;
+    };
+
+    // Cartesian base represented by unit vectors.
+    constexpr std::array<vector, spacedim> directions = {
+      {vector({1., 0., 0.}), vector({0., 1., 0.}), vector({0., 0., 1.})}};
+
+    // The skeleton corresponds to the axis of symmetry in the center of each
+    // pipe segment. Each skeleton vector points from the associated opening to
+    // the common bifurcation point. For convenience, we also compute length and
+    // unit vector of every skeleton vector here.
+    const auto skeleton = [&]() {
+      std::array<vector, n_pipes> skeleton;
```

Comment:
MSVC keeps complaining about this part.
```
D:\a\dealii\dealii\source\grid\grid_generator.cc(5149,1): error C2975: '_Size': invalid template argument for 'std::array', expected compile-time constant expression [D:\a\dealii\dealii\build\source\grid\obj_grid_debug.vcxproj]
```
`n_pipes` is constexpr and defined in the function body. I don't know what `MSVC` is complaining about. Do you have an idea? Otherwise I would get rid of `n_pipes` altogether which would introduce magic numbers...

Reply:
Seems to be an unfixed bug in MSVC since 2018. See https://developercommunity.visualstudio.com/t/invalid-template-argument-expected-compile-time-co/187862 and https://godbolt.org/z/sXDpFo which is a minimal example from the website.

Reply:
Just duplicate the declaration of `n_pipes` in the lambda function, and add a TODO about the duplication and its reason.